Source/WebCore/rendering/MarkedText.cpp

```diff
@@ -113,19 +113,38 @@ Vector<MarkedText> MarkedText::collectForHighlights(const RenderText& renderer,
         auto& parentRenderer = *renderer.parent();
         auto& parentStyle = parentRenderer.style();
         if (auto highlightRegister = renderer.document().highlightRegisterIfExists()) {
-            for (auto& highlight : highlightRegister->map()) {
-                auto renderStyle = parentRenderer.getUncachedPseudoStyle({ PseudoId::Highlight, highlight.key }, &parentStyle);
+            for (auto& highlightName : highlightRegister->highlightNames()) {
+                auto renderStyle = parentRenderer.getUncachedPseudoStyle({ PseudoId::Highlight, highlightName }, &parentStyle);
                 if (!renderStyle)
                     continue;
                 if (renderStyle->textDecorationsInEffect().isEmpty() && phase == PaintPhase::Decoration)
                     continue;
-                for (auto& rangeData : highlight.value->rangesData()) {
+                for (auto& rangeData : highlightRegister->map().get(highlightName)->rangesData()) {
                     if (!highlightData.setRenderRange(rangeData))
                         continue;
 
                     auto [highlightStart, highlightEnd] = highlightData.rangeForTextBox(renderer, selectableRange);
-                    if (highlightStart < highlightEnd)
-                        markedTexts.append({ highlightStart, highlightEnd, MarkedText::Type::Highlight, nullptr, highlight.key });
+
+                    if (highlightStart < highlightEnd) {
+                        int currentPriority = highlightRegister->map().get(highlightName)->priority();
+                        // If we can just append it to the end, do that instead.
+                        if (markedTexts.isEmpty() || markedTexts.last().priority <= currentPriority)
+                            markedTexts.append({ highlightStart, highlightEnd, MarkedText::Type::Highlight, nullptr, highlightName, currentPriority });
+                        else {
+                            // Find the first correct place to insert highlight.
+                            bool wasInserted = false;
+                            for (int index = markedTexts.size() - 1; index >= 0; index--) {
+                                if (markedTexts[index].priority <= currentPriority) {
+                                    markedTexts.insert(index + 1, { highlightStart, highlightEnd, MarkedText::Type::Highlight, nullptr, highlightName, currentPriority });
+                                    wasInserted = true;
+                                    break;
+                                }
+                            }
+                            // Insert at front of vector if lower priority than everything in markedTexts.
+                            if (!wasInserted)
```

Comment:
I really want this to be added into the above loop, but I'm not sure how to do it cleanly. But having the same insert line can lead to bugs in the future, so we should see if there's a nice way to combine these. But that can be in a follow up.

Reply:
Hmm... it seems like `markedTexts` is sorted in descending order.

If so, couldn't we do this?

```cpp
// Finds iterator of the first markedText with priority <= currentPriority
auto it = std::upper_bound(markedTexts.begin(), markedTexts.end(), key, [](const auto& m1, const auto& m2) {
    return m1.priority >= m2.priority();
});

unsigned insertIndex = (it == markedTexts.end() ? 0 : std::distance(markedTexts.begin(), it) + 1);
markedTexts.insert(insertIndex, { highlightStart, highlightEnd, MarkedText::Type::Highlight, nullptr, highlightName, currentPriority }
```

Reply:
above loop as in `for (auto& highlightName : highlightRegister->highlightNames())`?